### ISSUE TITLE
Output mismatch of torch.tan due to an intermediate result output when running on GPU

### ISSUE THREAD

**Issue by Azyka** [NVIDIA/TensorRT#3449]
## Description

When outputting an intermediate result of `torch.neg` in this model:
<img width="197" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/74590664/03b312e9-a11c-4345-8d79-847ee375128e">

New:
<img width="240" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/74590664/ee324dff-70d8-498e-a445-0827ae6d2e43">

The output of `torch.tan` is expected to be the same for the same input in this 2 graphs. However, it mismatched bewteew the 2 models.

## Environment

**TensorRT Version**: 8.6.1.post1

**NVIDIA GPU**: RTX 1660

**NVIDIA Driver Version**: 525.147.05

**CUDA Version**: 12.0

**CUDNN Version**: 8.9.4.25

Operating System: Ubuntu 22.04.3 LTS (x86_64)

Python Version (if applicable): 3.10.12

Tensorflow Version (if applicable): 2.13.0

PyTorch Version (if applicable): 2.1.0+cu118

## Relevant Files

**Model link**: 
[models.zip](https://github.com/NVIDIA/TensorRT/files/13335332/models.zip)

Input data file: 
[input_data.zip](https://github.com/NVIDIA/TensorRT/files/13335338/input_data.zip)

## Steps To Reproduce

**Script**:
```python
from dataclasses import dataclass
from numpy import testing
import numpy as np
import torch
import tensorrt as trt
import pycuda.driver as cuda
from pycuda.driver import DeviceAllocation
import pickle


@dataclass
class HostDeviceMem:
    host: np.ndarray
    device: DeviceAllocation


class ONNXClassifierWrapper():
    def __init__(self, engine):
        self.engine = engine

    def allocate_memory(self):
        engine = self.engine
        inputs = []
        outputs = []
        bindings = []
        stream = cuda.Stream()
        onames = []
        name2idx = {}
        for idx, binding in enumerate(engine):
            name2idx[binding] = idx
            size = trt.volume(engine.get_binding_shape(binding)) * engine.max_batch_size
            dtype = trt.nptype(engine.get_binding_dtype(binding))
            # Allocate host and device buffers
            host_mem = cuda.pagelocked_empty(size, dtype)
            device_mem = cuda.mem_alloc(host_mem.nbytes)
            # Append the device buffer to device bindings.
            bindings.append(int(device_mem))
            # Append to the appropriate list.
            if engine.binding_is_input(binding):
                inputs.append(HostDeviceMem(host_mem, device_mem))
            else:
                outputs.append(HostDeviceMem(host_mem, device_mem))
                onames.append(binding)
        return inputs, outputs, bindings, stream, onames, name2idx

    def predict(self, inputs):  # result gets copied into output
        (
            trt_inputs,
            trt_outputs,
            trt_bindings,
            stream,
            onames,
            name2idx,
        ) = self.allocate_memory()
        context = self.engine.create_execution_context()
        # print(name2idx)
        # print(inputs.keys())
        for iname in inputs:
            np.copyto(
                trt_inputs[name2idx[iname]].host,
                inputs[iname]
                .astype(trt.nptype(self.engine.get_binding_dtype(iname)))
                .ravel(),
            )

        [cuda.memcpy_htod_async(inp.device, inp.host, stream) for inp in trt_inputs]
        context.execute_async_v2(bindings=trt_bindings, stream_handle=stream.handle)
        [cuda.memcpy_dtoh_async(out.host, out.device, stream) for out in trt_outputs]
        stream.synchronize()
        trt_outputs = [out.host for out in trt_outputs]
        return {
            n: v.reshape(self.engine.get_binding_shape(n))
            for n, v in zip(onames, trt_outputs)
        }


def convert_onnx_to_engine(onnx_filename):
    logger = trt.Logger(trt.Logger.WARNING)
    with trt.Builder(logger) as builder, builder.create_network(
            1 << (int)(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)) as network, \
            trt.OnnxParser(network, logger) as parser:
        # builder.max_workspace_size = max_workspace_size
        # builder.fp16_mode = fp16_mode
        # builder.max_batch_size = max_batch_size

        config = builder.create_builder_config()
        config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, 2 << 30)

        with open(onnx_filename, 'rb') as model:
            if not parser.parse(model.read()):
                for error in range(parser.num_errors):
                    print(parser.get_error(error))

        engine_bytes = builder.build_serialized_network(network, config)

        return trt.Runtime(trt.Logger(trt.Logger.WARNING)).deserialize_cuda_engine(
            engine_bytes
        )

import pycuda.autoinit

class Model0(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, *args):
        _args = args
        getitem = _args[0];  _args = None
        sin = torch.sin(getitem);  getitem = None
        ceil = torch.ceil(sin)
        div = torch.div(ceil, ceil)
        return (div,)

model_0 = Model0()
output_names_0 = ['v3_0']
input_dict_0 = pickle.load(open('0.pickle', 'rb'))
inputs_0 = tuple(torch.from_numpy(v).to('cuda') for _, v in input_dict_0.items())
torch.onnx.export(model_0, inputs_0, '0.onnx', verbose=False, input_names=['v0_0'], output_names=output_names_0, opset_version=14, do_constant_folding=False)

class Model1(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, *args):
        _args = args
        getitem = _args[0]
        sin = torch.sin(getitem)
        ceil = torch.ceil(sin)
        div = torch.div(ceil, ceil)
        return (sin, div)

model_1 = Model1()
output_names_1 = ['v2_0', 'v3_0']
input_dict_1 = pickle.load(open('0.pickle', 'rb'))
inputs_1 = tuple(torch.from_numpy(v).to('cuda') for _, v in input_dict_1.items())
torch.onnx.export(model_1, inputs_1, '1.onnx', verbose=False, input_names=['v0_0'], output_names=output_names_1, opset_version=14, do_constant_folding=False)

engine_0 = convert_onnx_to_engine('0.onnx')
wrapper_0 = ONNXClassifierWrapper(engine_0)
output_0 = wrapper_0.predict(input_dict_0)

engine_1 = convert_onnx_to_engine('1.onnx')
wrapper_1 = ONNXClassifierWrapper(engine_1)
output_1 = wrapper_1.predict(input_dict_1)
output_name_dict = {'v3_0': 'v3_0'}

print('=========================')
try:
    for tensor_name_0, tensor_name_1 in output_name_dict.items():
        testing.assert_allclose(output_0[tensor_name_0], output_1[tensor_name_1], rtol=1, err_msg=f'at {tensor_name_0}, {tensor_name_1}')
    print("tensorRT does not trigger assertion")
except AssertionError as e:
    print("tensorRT triggers assertion")
    print(e)
print('=========================')
```

Steps to repro:
1. Download the input data file and put it at the same dir of the script.
2. Run the script

Output assertion:
```
=========================
tensorRT triggers assertion

Not equal to tolerance rtol=1, atol=0
at v3_0, v3_0
Mismatched elements: 17366 / 53312 (32.6%)
Max absolute difference: 1079.
Max relative difference: 517.5
 x: array([[[[[ 1.5342e+00, -4.3640e-02,  1.1963e+00, ..., -2.8735e-01,
            2.6680e+00, -1.7883e-01]]],
...
 y: array([[[[[-1.6934e+00,  1.1924e+00,  2.3867e+00, ..., -1.5533e-02,
            2.1895e+00, -1.7883e-01]]],
...
=========================
```

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**: No

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`): Yes

**Comment by zerollzeng** [comment]
I can reproduce the issue, file internal bug 4379250 to track this.

**Comment by zerollzeng** [comment]
A simplified code to reproduce:
```
$ cat compare.py 
import numpy as np
import pickle
from polygraphy.backend.onnxrt import OnnxrtRunner, SessionFromOnnx
from polygraphy.backend.trt import CreateConfig, EngineFromNetwork, NetworkFromOnnxPath, SaveEngine, TrtRunner
from polygraphy.comparator import Comparator, CompareFunc
from polygraphy.comparator import util as comp_util

def main():
    engine0 = EngineFromNetwork(NetworkFromOnnxPath("0.onnx"))
    engine1 = EngineFromNetwork(NetworkFromOnnxPath("1.onnx"))
    onnx_session0 = SessionFromOnnx("0.onnx")
    onnx_session1 = SessionFromOnnx("1.onnx")

    data = pickle.load(open('0.pickle', 'rb'))
    input_data = np.array(data["v0_0"], dtype=np.float16)
    comp_util.log_output_stats(input_data, None, f"input_data", hist_range=None)
    onnx_runners = [OnnxrtRunner(onnx_session0), OnnxrtRunner(onnx_session1)]
    onnx_results = Comparator.run(onnx_runners, data_loader=[{"v0_0":input_data}])
    Comparator.compare_accuracy(onnx_results, compare_func=CompareFunc.simple(atol=1e-6))
    runners = [TrtRunner(engine0), TrtRunner(engine1)]
    run_results = Comparator.run(runners, data_loader=[{"v0_0":input_data}])
    assert bool(Comparator.compare_accuracy(run_results, compare_func=CompareFunc.simple(atol=1e-6)))

    print("Test pass")

if __name__ == "__main__":
    main()
```

**Comment by zerollzeng** [comment]
This is not a bug, but a FP16 tolerance and nan issue.

**Comment by zerollzeng** [comment]
@Azyka We noticed you file several similar bug and most of them looks like corner case and seems not from real scenario(I can see you also file similar bug to other framework like pytorch/onnxruntime). May I ask what is your purpose? Are you doing some study on accuracy? Thanks!

**Comment by Azyka** [comment]
@zerollzeng We are conducting a study to identify potential bugs that may exist in the compilation optimization process of DL models. I'm sorry we didn't realize these were float accuracy issues. We simply noticed that there were discrepancies in the output results between TensorRT and other frameworks for these models. We apologize if these issues have caused any inconvenience to your work. If the issues we reported are confirmed to be accuracy issues, you just close them at will. Perhaps you could delve deeper into issues #3476 and #3477, as they seem not be related to float accuracy.

**Comment by zerollzeng** [comment]
Thanks! your issues are always welcome and I really appreciate your contribution :-) FYI it's a good practice to check input data range when checking accuracy issue. Also TRT doesn't provide bit-wise accuracy compare to onnxruntime/pytorch due to float-point difference. Using Polygraphy is a convenient way to check the diff range.

#3476 and #3477 will be fixed in TRT 10.

Closed this.
